projects/js-packages/components/components/alert/index.tsx

```diff
@@ -21,7 +20,17 @@ export const LEVEL_SUCCESS = 'success';
 
 export const ALERT_LEVELS = [ LEVEL_ERROR, LEVEL_WARNING, LEVEL_INFO, LEVEL_SUCCESS ];
```

Comment:
If you cast it `as const`, it will become a touple (stricter) instead of `string[]` and it will allow you simplify the type for `AlertProps['level']`, see the suggestion below.
```suggestion
export const ALERT_LEVELS = [ LEVEL_ERROR, LEVEL_WARNING, LEVEL_INFO, LEVEL_SUCCESS ] as const;
```

<img width="1099" alt="Screenshot 2022-05-02 at 5 53 52 PM" src="https://user-images.githubusercontent.com/18226415/166233009-f30290c5-d3f1-4b8c-8383-5f0b0cfb6ffe.png">

Reply:
@retrofox and @manzoorwanijk I would say to follow with `enum` instead of casting for `const`:

```
enum AlertLevels {
  ERROR = "error",
  WARNING = "warning",
  INFO = "info",
  SUCCESS = "success",
}
```

This makes more sense for the purpose of this one.

Reply:
It's better to avoid enums because they have size and runtime cost and the zero cost alternatives are better than enums.

Reply:
More details here - TypeScript: Avoid enums: p7H4VZ-3sK-p2

Reply:
If I had to do this conversion, I would not create those constants at all and I would rather use a Union of those string values which is pretty easy to work with and gives you much more flexibility in IDE. Also, I would use `React.FC` here because to make `defaultProps` typed automagically.

Here is what I would do...
```tsx
/**
 * External dependencies
 */
import React from 'react';
import { Icon, warning, info, check } from '@wordpress/icons';
import classNames from 'classnames';

/**
 * Internal dependencies
 */
import styles from './style.module.scss';

export type AlertLevel = 'error' | 'warning' | 'info' | 'success';

type AlertProps = {
	/** The severity of the alert. */
	level: AlertLevel;

	/** Show/Hide icon */
	showIcon?: boolean;
};

const getIconByLevel = ( level: AlertProps[ 'level' ] ) => {
	switch ( level ) {
		case 'error':
		case 'warning':
			return warning;
		case 'info':
			return info;
		case 'success':
			return check;
		default:
			return warning;
	}
};

/**
 * Alert component
 *
 * @param {AlertProps} props  - The component properties.
 * @returns {React.Component} The `Alert` component.
 */
const Alert: React.FC< React.PropsWithChildren< AlertProps > > = ( {
	level,
	children,
	showIcon,
} ) => {
	const classes = classNames( styles.container, styles[ `is-${ level }` ] );

	return (
		<div className={ classes }>
			{ showIcon && <Icon icon={ getIconByLevel( level ) } className={ styles.icon } /> }
			{ children }
		</div>
	);
};

Alert.defaultProps = {
	level: 'warning',
	showIcon: true,
};

export default Alert;
```

This is what would happen

<img width="365" alt="Screenshot 2022-05-02 at 9 19 24 PM" src="https://user-images.githubusercontent.com/18226415/166265166-d39afa14-89df-4fde-a8d8-cfcd9368d8de.png">
<img width="741" alt="Screenshot 2022-05-02 at 9 18 36 PM" src="https://user-images.githubusercontent.com/18226415/166265182-60684a6e-1307-4d93-a132-021fe1de85c6.png">

Reply:
The only thing that I could consider we export these constants to be consumed by other components.
Although it's true that TS will help us to set values for the level prop, it will work only for components written in TS. Meaning that, for instance, if we import and use the `<Alert />` by another no-written-in-ts component, we won't be able to prevent assigning a wrong value.
Not a bg deal at all, just saying.

Reply:
> if we import and use the `<Alert />` by another no-written-in-ts component, we won't be able to prevent assigning a wrong value.

That's a genuine concern 👍 

But, we can still import a type (e.g. `AlertLevel` I defined above) in a JS file and use JSDoc `@type {AlertLevel}` to set the correct value for a prop or a variable.

Reply:
💯

Reply:
> More details here - TypeScript: Avoid enums (by @sirreal): p7H4VZ-3sK-p2

Really good post. Thanks for sharing it. Let's follow with the first approach using cast `as const`.

Reply:
> I think the first approach works better at least with the IDE:

It's because I have extracted `AlertLevel` to its own type, if we don't do that (see the updated version), it should work just like the first approach.

Reply:
addressed [here](https://github.com/Automattic/jetpack/pull/24204/commits/5f2727f9e0a6a4be0b90ad87c8c2a56389c5904c).

I kept the constants there.